src/Evolution/DiscontinuousGalerkin/Actions/NormalCovectorAndMagnitude.hpp

```diff
@@ -0,0 +1,181 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <memory>
+#include <optional>
+#include <unordered_map>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/EagerMath/Magnitude.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Domain/CoordinateMaps/CoordinateMap.hpp"
+#include "Domain/CoordinateMaps/Tags.hpp"
+#include "Domain/Structure/Direction.hpp"
+#include "Evolution/DiscontinuousGalerkin/NormalVectorTags.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TypeTraits/CreateHasTypeAlias.hpp"
+
+namespace evolution::dg::Actions::detail {
+CREATE_HAS_TYPE_ALIAS(inverse_spatial_metric_tag)
+CREATE_HAS_TYPE_ALIAS_V(inverse_spatial_metric_tag)
+
+template <bool HasInverseSpatialMetricTag = false>
+struct inverse_spatial_metric_tag_impl {
+  template <typename System>
+  using f = tmpl::list<>;
+};
+
+template <>
+struct inverse_spatial_metric_tag_impl<true> {
+  template <typename System>
+  using f = tmpl::list<typename System::inverse_spatial_metric_tag>;
+};
+
+template <typename System>
+using inverse_spatial_metric_tag = typename inverse_spatial_metric_tag_impl<
+    has_inverse_spatial_metric_tag_v<System>>::template f<System>;
+
+struct OneOverNormalVectorMagnitude : db::SimpleTag {
+  using type = Scalar<DataVector>;
+};
+
+template <size_t Dim>
+struct NormalVector : db::SimpleTag {
+  using type = tnsr::I<DataVector, Dim, Frame::Inertial>;
+};
+
+/*!
+ * \brief Computes the normal covector, magnitude of the unnormalized normal
+ * covector, and in curved spacetimes the normal vector.
```

Comment:
It is all three. I'm not sure how much is improved by factoring because I sort of need the inverse spatial metric sometimes. I really don't want to have to function calls, one for with and one for without the spatial metric. Doing so would make the ComputeTimeDerivative action even more complicated, and I'd rather avoid that. The `NormalVector` and `OneOver...` are always returned and so need to be `not_null` anyway. Factoring these would make the call site more complicated (the TimeDerivativeAction) because then it's just `get<Tag>(fields_on_face), get<Tag>(fileds_on_face)`. Since this is effectively an `impl` function itself, I personally prefer the current `fields_on_face` part of the interface. I have added documentation to clarify what tags are needed and how they are used.